src/coprocessor/cache.rs

```diff
@@ -0,0 +1,490 @@
+        };
+        self.regions.insert(region_id, entry);
+    }
+
+    pub fn evict_region(&mut self, region_id: u64) {
+        if self.regions.contains_key(&region_id) {
+            self.evict_region_with_exist_region(region_id);
+        } else {
+            self.evict_region_with_new_region(region_id);
+        }
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    pub fn evict_region_and_enable(&mut self, region_id: u64) {
+        self.evict_region(region_id);
+        self.enable_region_cache(region_id);
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.max_size
+    }
+
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.map.len() == 0
+    }
+
+    pub fn update_capacity(&mut self, capacity: usize) {
+        self.max_size = capacity;
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+    }
+
+    pub fn disable_region_cache(&mut self, region_id: u64) {
+        let opt = match self.regions.get_mut(&region_id) {
+            Some(entry) => {
+                entry.enable = false;
+                return;
+            }
+            None => {
+                let rmap = HashMap::new();
+                Some(rmap)
+            }
+        };
+        if let Some(rmap) = opt {
+            let entry = RegionDistSQLCacheEntry {
+                version: 0,
+                enable: false,
+                cached_items: rmap,
+            };
+            self.regions.insert(region_id, entry);
+        }
+    }
+
+    pub fn enable_region_cache(&mut self, region_id: u64) {
+        if let Some(entry) = self.regions.get_mut(&region_id) {
+            entry.enable = true;
+        }
+    }
+
+    fn is_region_cache_enabled(&self, region_id: u64) -> bool {
+        match self.regions.get(&region_id) {
+            None => true,
+            Some(entry) => entry.enable,
+        }
+    }
+
+    fn update_regions(&mut self, region_id: u64, k: DistSQLCacheKey) {
+        let opt = match self.regions.get_mut(&region_id) {
+            Some(entry) => {
+                entry.cached_items.insert(k, 1);
+                None
+            }
+            None => {
+                let mut rmap = HashMap::new();
+                rmap.insert(k, 1);
+                Some(rmap)
+            }
+        };
+        if let Some(rmap) = opt {
+            let entry = RegionDistSQLCacheEntry {
+                version: 0,
+                enable: true,
+                cached_items: rmap,
+            };
+            self.regions.insert(region_id, entry);
+        }
```

Comment:
Maybe you could use `entry(xxx).or_insert(xxx)` instead (from L287-L305) like:

```
let mut opt = self.regions.entry(region_id).or_insert(RegionDistSQLCacheEntry {
                enable: true,
                version: 0,
                cached_items:HashMap::default()
            });
        opt.cached_items.insert(k,1);
```

And Could we implement  `Default` for `RegionDistSQLCacheEntry` with enable=true? Then this function would be more  easier like:
```
let mut opt = self.regions.entry(region_id).or_insert(Default::default());
opt.cached_items.insert(k,1);
```